testcases/open_hpi_testsuite/docs/man/hpiel.pod.1
=head1 NAME

hpiel - An openhpi sample application that displays HPI event log entries. 

=head1 SYNOPSIS 

hpiel [ -d -e -c -p -r -x -v -h ]

=head1 DESCRIPTION

hpiel searches the RPT (Resource Present Table) for resources with SAHPI_CAPABILITY_EVENT_LOG and displays event log entries for resources found.

=head1 OPTIONS

=over 2

 --del, -d				display domain event log entries 
 --entitypath="<arg>", -e "<arg>"	display resource event log entries
                                        (e.g. -e "{SYSTEM_CHASSIS,2}{SBC_BLADE,5}")
 --clear, -c	clear log before reading event log entries
 --resource, -p	pull resource info along with log entry
 --rdr, -r	pull RDR info along with log entry
 --xml, -x	print output in xml format (not implemented)
 --verbose, -v	print debug messages
 --help, -h	print this usage message

 If neither -d or -e "<arg>" are specified, event log entries will be shown
for all supporting resources by default.

=back

=head1 HPI APIs uniquely used in this application 

=over 2

SaErrorT SAHPI_API saHpiEventLogInfoGet (
    SAHPI_IN  SaHpiSessionIdT     SessionId,
    SAHPI_IN  SaHpiResourceIdT    ResourceId,
    SAHPI_OUT SaHpiEventLogInfoT  *Info
);

SaErrorT SAHPI_API saHpiEventLogEntryGet (
    SAHPI_IN    SaHpiSessionIdT          SessionId,
    SAHPI_IN    SaHpiResourceIdT         ResourceId,
    SAHPI_IN    SaHpiEventLogEntryIdT    EntryId,
    SAHPI_OUT   SaHpiEventLogEntryIdT    *PrevEntryId,
    SAHPI_OUT   SaHpiEventLogEntryIdT    *NextEntryId,
    SAHPI_OUT   SaHpiEventLogEntryT      *EventLogEntry,
    SAHPI_INOUT SaHpiRdrT                *Rdr,
    SAHPI_INOUT SaHpiRptEntryT           *RptEntry
);

SaErrorT SAHPI_API saHpiEventLogClear (
    SAHPI_IN  SaHpiSessionIdT     SessionId,
    SAHPI_IN  SaHpiResourceIdT    ResourceId
);

=back

=head1 SEE ALSO

=head1 AUTHORS

Authors of this man page:

 Peter D Phan (pdphan@users.sourceforge.net)



ChangeLog
LTP-20081031

1) Log Message:
Here is one fix for io-controller testcase. It might not have run when kernel supports io-controller. Description: Instead of block io testcases, memory controller testcase will run. Signed-off By: Rishikesh K Rajak <risrajak@linux.vnet.ibm.com>.

Modified File(s):
ltp/testcases/kernel/controllers/test_controllers.sh

2) Log Message:
test_ioctl produces a 'Syntax error: Bad substitution: 28' under Ubuntu (and much probably Debian). The following patch solves this issue. I tested it with sh (aka dash) and bash. Signed-Off-By: Elder Costa <elder.costa@terra.com.br>.

Modified File(s):
ltp/testcases/kernel/syscalls/ioctl/test_ioctl

3) Log Message:
Add support of kernel syscall test to Renesas. Signed-off-by: Nobuhiro Iwamatsu <iwamatsu.nobuhiro@renesas.com>.

Modified Files:
ltp/testcases/kernel/include/order
Added Files:
ltp/testcases/kernel/include/sh.in

4) Log Message:
Reverting the Following Patch.
test_ioctl produces a 'Syntax error: Bad substitution: 28' under Ubuntu (and much probably Debian). The following patch solves this issue. I tested it with sh (aka dash) and bash. Signed-Off-By: Elder Costa <elder.costa@terra.com.br>.

Modified File(s):
ltp/testcases/kernel/syscalls/ioctl/test_ioctl

5) Log Message:
The clock source used in the gtod_infinite test should be CLOCK_MONOTONIC instead of CLOCK_REALTIME. This is to avoid plausible inconsistencies caused by NTP time adjustments. Tested: running gtod_infinite and then killing it. Signed-Off-By: Chirag <chirag@linux.vnet.ibm.com>, Acked-by: Darren Hart <dvhltc@us.ibm.com>.

Modified File(s):
ltp/testcases/realtime/func/gtod_latency/gtod_infinite.c

6) Log Message:
Hi Andrea, As you are using myfunctions.sh in your script, it will simply replace the installed myfunctions.sh in $LTP_ROOT/testcase/bin which is written by Sudhir. And if you will run this testcase from `./runltp` sudhir's memory controller tetscase will refer your myfuncitons.sh script. So i am just renaming this script to some other name and correspondingly changing the Makefile & run_io_throttle_test.sh . Attached is the patch. I might require your acknowledgement for this. Signed-Off-By: "Rishikesh K. Rajak" <risrajak@linux.vnet.ibm.com>, Acked-by: Andrea Righi <righi.andrea@gmail.com>.

Modified Files:
ltp/testcases/kernel/controllers/io-throttle/Makefile
ltp/testcases/kernel/controllers/io-throttle/run_io_throttle_test.sh
Added Files:
ltp/testcases/kernel/controllers/io-throttle/myfunctions-io.sh
Removed Files:
ltp/testcases/kernel/controllers/io-throttle/myfunctions.sh

7) Log Message:
dont use == or [[, and fix -z test

Modified File(s):
ltp/testcases/network/tcp_cmds/tcpdump/tcpdump01

8) Log Message:
convert all "source" commands in scripts to "." ... hopefully my sed-foo doesnt break anything ...

Modified File(s):
ltp/testcases/network/stress/broken_ip/broken_ip4-checksum
ltp/testcases/network/stress/broken_ip/broken_ip4-dstaddr
ltp/testcases/network/stress/broken_ip/broken_ip4-fragment
ltp/testcases/network/stress/broken_ip/broken_ip4-ihl
ltp/testcases/network/stress/broken_ip/broken_ip4-protcol
ltp/testcases/network/stress/broken_ip/broken_ip4-totlen
ltp/testcases/network/stress/broken_ip/broken_ip4-version
ltp/testcases/network/stress/broken_ip/broken_ip6-dstaddr
ltp/testcases/network/stress/broken_ip/broken_ip6-nexthdr
ltp/testcases/network/stress/broken_ip/broken_ip6-plen
ltp/testcases/network/stress/broken_ip/broken_ip6-version
ltp/testcases/network/stress/ftp/ftp4-download-stress
ltp/testcases/network/stress/ftp/ftp4-upload-stress
ltp/testcases/network/stress/ftp/ftp6-download-stress
ltp/testcases/network/stress/ftp/ftp6-upload-stress
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic01
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic02
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic03
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic04
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic05
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic06
ltp/testcases/network/stress/icmp/uni-basic/icmp4-uni-basic07
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic01
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic02
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic03
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic04
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic05
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic06
ltp/testcases/network/stress/icmp/uni-basic/icmp6-uni-basic07
ltp/testcases/network/stress/interface/if4-addr-change
ltp/testcases/network/stress/interface/if4-alias-adddel
ltp/testcases/network/stress/interface/if4-alias-addlarge
ltp/testcases/network/stress/interface/if4-mtu-change
ltp/testcases/network/stress/interface/if4-route-adddel
ltp/testcases/network/stress/interface/if4-route-addlarge
ltp/testcases/network/stress/interface/if4-updown
ltp/testcases/network/stress/interface/if6-addr-adddel
ltp/testcases/network/stress/interface/if6-addr-addlarge
ltp/testcases/network/stress/interface/if6-mtu-change
ltp/testcases/network/stress/interface/if6-route-adddel
ltp/testcases/network/stress/interface/if6-route-addlarge
ltp/testcases/network/stress/interface/if6-updown
ltp/testcases/network/stress/http/http4-stress
ltp/testcases/network/stress/http/http6-stress
ltp/testcases/network/stress/multicast/grp-operation/mcast4-grpope01
ltp/testcases/network/stress/multicast/grp-operation/mcast4-grpope02
ltp/testcases/network/stress/multicast/grp-operation/mcast4-grpope03
ltp/testcases/network/stress/multicast/grp-operation/mcast4-grpope04
ltp/testcases/network/stress/multicast/grp-operation/mcast6-grpope01
ltp/testcases/network/stress/multicast/grp-operation/mcast6-grpope02
ltp/testcases/network/stress/multicast/grp-operation/mcast6-grpope03
ltp/testcases/network/stress/multicast/grp-operation/mcast6-grpope04
ltp/testcases/network/stress/multicast/packet-flood/mcast4-pktfld01

testcases/open_hpi_testsuite/clients/hpiel.c
/* -*- linux-c -*-
 *
 * (C) Copyright IBM Corp 2006
 *
 * Author(s):
 *     Renier Morales <renier@openhpi.org>
 *
 * hpiel - Displays HPI event log entries.
 *
 */

#include <stdlib.h>
#include <string.h>
#include <getopt.h>


#include <SaHpi.h>
#include <oHpi.h>
#include <oh_utils.h>
#include <oh_clients.h>

#define OH_SVN_REV "$Revision: 1.8 $"

#define err(format, ...) \
        do { \
                if (opts.dbg) { \
                        fprintf(stderr, format "\n", ## __VA_ARGS__); \
                } \
        } while(0)

#define show_error_quit(msg) \
        do { \
                if (error) { \
			opts.dbg = 1; \
                        err(msg, oh_lookup_error(error)); \
                        exit(-1); \
                } \
        } while(0)

/* Globals */
static struct hpiel_opts {
        int  del;        /* Domain Event Log option. */
        char *ep;       /* Resource Event Log option. */
        int  clear;      /* Clear the event log before traversing it. */
        int  resource;   /* Get resource along with event log entry. */
        int  rdr;        /* Get RDR along with event log entry. */
        int  dbg;        /* Display debug messages. */
} opts = { 0, NULL, 0, 0, 0, 0 };

/* Prototypes */
SaErrorT parse_options(int argc, char ***argv, struct hpiel_opts *opts);
SaErrorT harvest_sels(SaHpiSessionIdT sid, SaHpiDomainInfoT *dinfo, char *ep);
SaErrorT display_el(SaHpiSessionIdT sid, SaHpiResourceIdT rid, SaHpiTextBufferT *tag);

int main(int argc, char **argv)
{
        SaErrorT error = SA_OK;
        SaHpiSessionIdT sid;
        SaHpiDomainInfoT dinfo;

        /* Print version strings */
	oh_prog_version(argv[0], OH_SVN_REV);

        /* Parsing options */
        if (parse_options(argc, &argv, &opts)) {
                fprintf(stderr, "There was an error parsing the options. Exiting.\n");
                exit(-1);
        }

        /* Program really begins here - all options parsed at this point */
        error = saHpiSessionOpen(SAHPI_UNSPECIFIED_DOMAIN_ID, &sid, NULL);
        show_error_quit("saHpiSessionOpen() returned %s. Exiting.\n");

        error = saHpiDiscover(sid);
        show_error_quit("saHpiDiscover() returned %s. Exiting.\n");

        error = saHpiDomainInfoGet(sid, &dinfo);
        show_error_quit("saHpiDomainInfoGet() returned %s. Exiting.\n");

        printf("Domain Info: UpdateCount = %d, UpdateTime = %lx\n",
               dinfo.RptUpdateCount, (unsigned long)dinfo.RptUpdateTimestamp);

        if (opts.ep) { /* Entity path specified */
                error = harvest_sels(sid, &dinfo, opts.ep);
        } else if (opts.del) { /* Domain event log specified */
                error = display_el(sid, SAHPI_UNSPECIFIED_RESOURCE_ID, &dinfo.DomainTag);
        } else { /* Else, show SELs of all supporting resources */
                error = harvest_sels(sid, &dinfo, NULL);
        }

        if (error) err("An error happened. Gathering event log entries returned %s",
                       oh_lookup_error(error));

        error = saHpiSessionClose(sid);
        if (error) err("saHpiSessionClose() returned %s.",
                       oh_lookup_error(error));

        return error;
}

#define print_usage_quit() \
        do { \
                printf("\nUsage:\n" \
                       "  hpiel - Displays HPI event log entries.\n\n" \
                       "    --del, -d                        display domain event log entries\n" \
                       "    --entitypath=\"<arg>\", -e \"<arg>\"     display resource event log entries\n" \
                       "            (e.g. -e \"{SYSTEM_CHASSIS,2}{SBC_BLADE,5}\")\n" \
                       "    --clear, -c                      clear log before reading event log entries\n" \
                       "    --resource, -p                   pull resource info along with log entry\n" \
                       "    --rdr, -r                        pull RDR info along with log entry\n" \
                       "    --xml, -x                        print output in xml format (not implemented)\n" \
                       "    --verbose, -v                    print debug messages\n" \
                       "    --help, -h                       print this usage message\n" \
                       "\n    If neither -d or -e \"<arg>\" are specified, event log entries will be shown\n" \
                       "    for all supporting resources by default.\n\n"); \
                exit(0); \
        } while(0)

SaErrorT parse_options(int argc, char ***argv, struct hpiel_opts *opts)
{
        static struct option long_options[] = {
                {"del",        no_argument,       0, 'd'},
                {"entitypath", required_argument, 0, 'e'},
                {"clear",      no_argument,       0, 'c'},
                {"resource",   no_argument,       0, 'p'},
                {"rdr",        no_argument,       0, 'r'},
                {"verbose",    no_argument,       0, 'v'},
                {"help",       no_argument,       0, 'h'},
                {0, 0, 0, 0}
        };
        int c, option_index = 0;

        while ((c = getopt_long(argc, *argv, "de:cprvh", long_options, &option_index)) != -1) {
                switch(c) {
                        case 'd':
                                opts->del = 1;
                                break;
                        case 'e':
                                opts->ep = strdup(optarg);
                                break;
                        case 'c':
                                opts->clear = 1;
                                break;
                        case 'p':
                                opts->resource = 1;
                                break;
                        case 'r':
                                opts->rdr = 1;
                                break;
                        case 'v':
                                opts->dbg = 1;
                                break;
                        case 'h':
                                print_usage_quit();
                                break;
                        case '?':
                                printf("extraneous option found, %d\n", optopt);
                                break;
                        default:
                                printf("Found bad option %d.\n", c);
                                return -1;
                }
        }

        return 0;
}

SaErrorT harvest_sels(SaHpiSessionIdT sid, SaHpiDomainInfoT *dinfo, char *ep)
{
        SaErrorT error = SA_OK;
        SaHpiRptEntryT rptentry;
        SaHpiEntryIdT entryid, nextentryid;
        SaHpiResourceIdT rid;
        oh_big_textbuffer bigbuf;
        SaHpiEntityPathT entitypath;
        SaHpiBoolT found_entry = SAHPI_FALSE;

        if (!sid || !dinfo) {
                err("Invalid parameters in havest_sels()\n");
                return SA_ERR_HPI_INVALID_PARAMS;
        }

        if (opts.ep && ep) {
                error = oh_encode_entitypath(ep, &entitypath);
                if (error) {
                        err("oh_encode_entitypath() returned %s from %s\n",
                            oh_lookup_error(error), ep);
                        return error;
                }
        }

        entryid = SAHPI_FIRST_ENTRY;
        while (error == SA_OK && entryid != SAHPI_LAST_ENTRY) {
                error = saHpiRptEntryGet(sid, entryid, &nextentryid, &rptentry);

                err("saHpiRptEntryGet() returned %s\n", oh_lookup_error(error));
                if (error == SA_OK) {
                        if (opts.ep && ep) {
                                if (!oh_cmp_ep(&entitypath, &rptentry.ResourceEntity)) {
                                        entryid = nextentryid;
                                        continue;
                                }
                        }

                        if (!(rptentry.ResourceCapabilities & SAHPI_CAPABILITY_EVENT_LOG)) {
                                err("RPT doesn't have SEL\n");
                                entryid = nextentryid;
                                continue;  /* no SEL here, try next RPT */
                        }
                        found_entry = SAHPI_TRUE;

                        rid = rptentry.ResourceId;
                        err("RPT %d capabilities = %x\n",
                            rid, rptentry.ResourceCapabilities);
                        rptentry.ResourceTag.Data[rptentry.ResourceTag.DataLength] = 0;

                        oh_init_bigtext(&bigbuf);
                        error = oh_decode_entitypath(&rptentry.ResourceEntity, &bigbuf);
                        printf("%s\n", bigbuf.Data);
                        printf("rptentry[%d] tag: %s\n", rid, rptentry.ResourceTag.Data);

                        error = display_el(sid, rid, &rptentry.ResourceTag);

                        if (opts.ep && ep) return SA_OK;
                }

                entryid = nextentryid;
        }

        if (!found_entry) {
                if (opts.ep && ep) {
                        err("Could not find resource matching %s\n", ep);
                } else {
                        err("No resources supporting event logs were found.\n");
                }
        }

        return error;
}

SaErrorT display_el(SaHpiSessionIdT sid, SaHpiResourceIdT rid, SaHpiTextBufferT *tag)
{
        SaErrorT error = SA_OK;
        SaHpiEventLogEntryIdT entryid, nextentryid, preventryid;
        SaHpiEventLogInfoT elinfo;
        SaHpiEventLogEntryT elentry;
        SaHpiRdrT rdr;
        SaHpiRptEntryT res;

        if (!sid || !rid) {
                err("Invalid parameters in display_el()\n");
                return SA_ERR_HPI_INVALID_PARAMS;
        }

        error = saHpiEventLogInfoGet(sid, rid, &elinfo);
        if (error) {
                err("saHpiEventLogInfoGet() returned %s. Exiting\n",
                    oh_lookup_error(error));
                return error;
        }

        printf("EventLogInfo for %s, ResourceId %d\n",
               tag->Data, rid);
        oh_print_eventloginfo(&elinfo, 4);

        if (elinfo.Entries == 0) {
                printf("%s Resource %d has an empty event log.\n", tag->Data, rid);
                return SA_OK;
        }

        if (opts.clear) {
                error = saHpiEventLogClear(sid, rid);
                if (error == SA_OK)
                        printf("EventLog successfully cleared\n");
                else {
                        printf("saHpiEventLogClear() returned %s\n",
                               oh_lookup_error(error));
                        return error;
                }

        }

        entryid = SAHPI_OLDEST_ENTRY;
        while (entryid != SAHPI_NO_MORE_ENTRIES) {
                error = saHpiEventLogEntryGet(sid, rid,
                                              entryid, &preventryid,
                                              &nextentryid, &elentry,
                                              &rdr,
                                              &res);

                err("saHpiEventLogEntryGet() returned %s\n", oh_lookup_error(error));
                if (error == SA_OK) {
                	SaHpiEntityPathT *ep = NULL;
                	/* Get a reference to the entity path for this log entry */
                	if (res.ResourceCapabilities) {
                		ep = &res.ResourceEntity;
                	} else if (rdr.RdrType != SAHPI_NO_RECORD) {
                		ep = &rdr.Entity;
                	}
                	/* Print the event log entry */
                        oh_print_eventlogentry(&elentry, ep, 6);
                        if (opts.rdr) {
                                if (rdr.RdrType == SAHPI_NO_RECORD)
                                        printf("            No RDR associated with EventType =  %s\n\n",
                                               oh_lookup_eventtype(elentry.Event.EventType));
                                else
                                        oh_print_rdr(&rdr, 12);
                        }

                        if (opts.resource) {
                                if (res.ResourceCapabilities == 0)
                                        printf("            No RPT associated with EventType =  %s\n\n",
                                               oh_lookup_eventtype(elentry.Event.EventType));
                                else
                                        oh_print_rptentry(&res, 10);
                        }

                        preventryid = entryid;
                        entryid = nextentryid;
                } else {
                        return error;
                }
        }

        return SA_OK;
}
